File: java/src/main/java/helptree/examples/Hidden.java
package helptree.examples;

import helptree.HelpTree;
import helptree.HelpTree.Config;
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

@Command(name = "hidden", description = "An example with hidden commands and flags")
public class Hidden implements Runnable {

    @Option(names = {"--help-tree"}, description = "Print a recursive command map derived from framework metadata")
    boolean helpTree;

    @Option(names = {"-L", "--tree-depth"}, description = "Limit --help-tree recursion depth (Unix tree -L style)")
    Integer treeDepth;

    @Option(names = {"-I", "--tree-ignore"}, description = "Exclude subtrees/commands from --help-tree output (repeatable)")
    String[] treeIgnore;

    @Option(names = {"-a", "--tree-all"}, description = "Include hidden subcommands in --help-tree output")
    boolean treeAll;

    @Option(names = {"--tree-output"}, description = "Output format (text or json)")
    String treeOutput;

    @Option(names = {"--tree-style"}, description = "Tree text styling mode (rich or plain)")
    String treeStyle;

    @Option(names = {"--tree-color"}, description = "Tree color mode (auto, always, never)")
    String treeColor;

    @Option(names = {"--verbose"}, description = "Verbose output")
    boolean verbose;

    @Option(names = {"--debug"}, description = "Enable debug mode", hidden = true)
    boolean debug;

    @Command(name = "list", description = "List items")
    static class ListCmd implements Runnable {
        public void run() {}
    }

    @Command(name = "show", description = "Show item details")
    static class ShowCmd implements Runnable {
        @Parameters(paramLabel = "ID", description = "Item ID")
        String id;
        public void run() {}
    }

    @Command(name = "admin", description = "Administrative commands", hidden = true)
    static class AdminCmd implements Runnable {
        @Command(name = "users", description = "List all users")
        static class UsersCmd implements Runnable { public void run() {} }
        @Command(name = "stats", description = "Show system stats")
        static class StatsCmd implements Runnable { public void run() {} }
        @Command(name = "secret", description = "Secret backdoor")
        static class SecretCmd implements Runnable { public void run() {} }
        public void run() {}
    }

    public void run() {}

    public static void main(String[] args) {
        Hidden hidden = new Hidden();
        CommandLine cmd = new CommandLine(hidden);
        cmd.addSubcommand("list", new ListCmd());
        cmd.addSubcommand("show", new ShowCmd());
        CommandLine admin = new CommandLine(new AdminCmd());
        admin.addSubcommand("users", new AdminCmd.UsersCmd());
        admin.addSubcommand("stats", new AdminCmd.StatsCmd());
        admin.addSubcommand("secret", new AdminCmd.SecretCmd());
        cmd.addSubcommand("admin", admin);

        Config config = HelpTree.extractConfig(args);
        if (config.helpTree) {
            var tree = HelpTree.fromPicocli(cmd);
            tree = HelpTree.resolvePath(tree, config.path);
            System.out.println(HelpTree.render(tree, config));
            return;
        }

        int exitCode = cmd.execute(config.remainingArgs.toArray(new String[0]));
        System.exit(exitCode);
    }
}

File: kotlin/src/main/kotlin/helptree/examples/Hidden.kt
package helptree.examples

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.main
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.options.*
import helptree.*

class Hidden : CliktCommand("hidden") {
    val verbose by option("--verbose", help = "Verbose output").flag()
    val debug by option("--debug", help = "Enable debug mode", hidden = true).flag()
    override fun run() {}

    class List : CliktCommand("list") {
        override fun run() {}
    }
    class Show : CliktCommand("show") {
        val id by argument("ID", help = "Item ID")
        override fun run() {}
    }
    class Admin : CliktCommand("admin") {
        override val hiddenFromHelp = true
        override fun run() {}
        class Users : CliktCommand("users") { override fun run() {} }
        class Stats : CliktCommand("stats") { override fun run() {} }
        class Secret : CliktCommand("secret") { override fun run() {} }
    }
}

fun main(args: Array<String>) {
    System.setOut(java.io.PrintStream(System.out, true, java.nio.charset.StandardCharsets.UTF_8))

    val cfg = extractConfig(args)
    if (cfg.helpTree) {
        val verboseOpt = TreeOption("verbose", "", "--verbose", "Verbose output", required = false, takesValue = false)
        val debugOpt = TreeOption("debug", "", "--debug", "Enable debug mode", required = false, takesValue = false, hidden = true)
        val root = TreeCommand(
            name = "hidden",
            description = "An example with hidden commands and flags",
            options = discoveryOptions() + verboseOpt + debugOpt,
            subcommands = listOf(
                TreeCommand(name = "list", description = "List items"),
                TreeCommand(name = "show", description = "Show item details",
                    arguments = listOf(TreeArgument("ID", "Item ID", required = true))),
                TreeCommand(
                    name = "admin",
                    description = "Administrative commands",
                    hidden = true,
                    subcommands = listOf(
                        TreeCommand(name = "users", description = "List all users"),
                        TreeCommand(name = "stats", description = "Show system stats"),
                        TreeCommand(name = "secret", description = "Secret backdoor")
                    )
                )
            )
        )
        val selected = resolvePath(root, cfg.path)
        println(render(selected, cfg))
        return
    }

    Hidden().subcommands(
        Hidden.List(), Hidden.Show(),
        Hidden.Admin().subcommands(Hidden.Admin.Users(), Hidden.Admin.Stats(), Hidden.Admin.Secret())
    ).main(cfg.remainingArgs.toTypedArray())
}
